lgc/elfLinker/ElfLinker.cpp

```diff
@@ -234,27 +251,48 @@ inline uint64_t cantFail(uint64_t value, const char *Msg = nullptr) {
 // @param elfs : Array of unlinked ELF modules to link
 ElfLinkerImpl::ElfLinkerImpl(PipelineState *pipelineState, ArrayRef<MemoryBufferRef> elfs)
     : m_pipelineState(pipelineState), m_relocHandler(pipelineState) {
-  // For each input ELF, create an ELF object for it.
-  for (auto elfBuffer : elfs)
-    m_elfInputs.push_back({cantFail(object::ObjectFile::createELFObjectFile(elfBuffer))});
-
-  // Gather and merge PAL metadata.
   m_pipelineState->clearPalMetadata();
-  for (auto &elfInput : m_elfInputs)
-    mergePalMetadataFromElf(*elfInput.objectFile, false);
 
-  // Create any needed glue shaders.
-  createGlueShaders();
-
-  // Populate output ELF header
-  memcpy(&m_ehdr, elfs[0].getBuffer().data(), sizeof(ELF::Elf64_Ehdr));
+  // Add ELF inputs supplied here.
+  for (MemoryBufferRef elf : elfs)
+    addInputElf(elf, /*addAtStart=*/false);
 }
 
 // =====================================================================================================================
 // Destructor
 ElfLinkerImpl::~ElfLinkerImpl() {
 }
 
+// =====================================================================================================================
+// Add another input ELF to the link.
+void ElfLinkerImpl::addInputElf(MemoryBufferRef inputElf, bool addAtStart) {
+  assert(!m_doneInputs && "Cannot use ElfLinker::addInputElf after other ElfLinker calls");
+  // Populate output ELF header if this is the first one to be added.
+  if (m_elfInputs.empty())
+    memcpy(&m_ehdr, inputElf.getBuffer().data(), sizeof(ELF::Elf64_Ehdr));
```

Comment:
Will the buffer always be of sufficient size?

Reply:
Good point. I moved the createELFObjectFile up above that, so the cantFail will assert in that case.